### ISSUE TITLE
Polywrap cachelib - Niraj

### ISSUE THREAD

**Issue by Niraj-Kamdar** [polywrap/mini-hacks#26]
## Summary
A general interface for polywrap cache and a plugin for simple caching

## Description
Currently polywrappers are stateless and there isn't any way to store and retrieve state to/from client. Using this general cachelib interface user can store simple key-value data to any service (Ex: in-memory, filesystem, memcache, redis, localstorage, key-value db, etc). 

> Note: currently only simple in-memory cache has been implemented. 

## Deliverables
GitHub: https://github.com/Niraj-Kamdar/polywrap-cachelib

## Any next steps?
- Add more methods like batch operations (add, get, update, remove), counter (increment, decrement, etc). 
- Add more plugins or polywrappers that implements cache interface for different services. 

## About you
Polywrap and defiwrapper core dev

**Comment by dOrgJelli** [comment]
Bravo for finally creating this!

I really like the setup where you use an interface for the plugin. Doing this allows wrapper developers to depend on your plugin!

Here are the details for how this works (as I realize this isn't documented anywhere but it's super crucial it seems):
1. Plugin developer `implements` an interface within their schema, as you've done.
```graphql
#import { Query, Mutation } into Cache from "ens/interface.cachelib.eth"

type Query implements Cache_Query { }

type Mutation implements Cache_Mutation { }
```
2. The plugin manifest that gets generated when you run `w3 plugin codegen` contains all implemented interfaces.
```
// @ts-noCheck
import { schema } from "./";

import { PluginPackageManifest, Uri } from "@web3api/core-js";

export const manifest: PluginPackageManifest = {
  schema,
  implements: [
    new Uri("ens/interface.cachelib.eth")
  ],
}
```
3. The wrapper developer imports the interface as a dependency.
```graphql
#import { Query, Mutation } into Cache from "ens/interface.cachelib.eth"

type Mutation {
  interactWithCache(...): String!
}
```
4. Within the wrapper developers code, they'll query all implementations of the interface. We can make this easier to do, so that wrapper developers don't have to write the loop themselves (need to make an issue for this).
5. Within the dapp developer's code, they'll simply add the plugin to their client config as normal, and when the invoke the wrapper, the plugin's implementation will get picked up via the manifest shown above.

**Comment by dOrgJelli** [comment]
Very excited to leverage this to implement client-side Polywrap caching! https://github.com/polywrap/monorepo/issues/560

**Comment by Niraj-Kamdar** [comment]
Yeah, I agree on all the points and we should have an interface for all the plugins. We only need URI for the interface. Plugins are local dependency and shouldn't have `ens` authority URI, it can have either: `local`, `fs`, or `npm` as authority. 

Regarding 2nd point, I believe, It will allow us to use interfaces for plugins without needing to explicitly write it in client configs, Correct me if I am wrong.
